File: pegasusX/apps/warehouse-app-android/app/src/main/java/com/pegasusx/warehouse/data/model/WarehouseOpsModels.kt
package com.pegasusx.warehouse.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class EmergencyTransferRequest(
    @SerialName("total_volume_vu") val totalVolumeVu: Double,
    val notes: String? = null,
)

@Serializable
data class ForceReceiveRequest(
    @SerialName("factory_id") val factoryId: String? = null,
    @SerialName("total_volume_vu") val totalVolumeVu: Double,
    val notes: String? = null,
)

@Serializable
data class TransferMutationResponse(
    @SerialName("transfer_id") val transferId: String,
    val state: String,
    val notes: String? = null,
)

@Serializable
data class ReplenishmentInsight(
    val id: String,
    @SerialName("warehouse_id") val warehouseId: String,
    @SerialName("warehouse_name") val warehouseName: String,
    @SerialName("product_id") val productId: String,
    @SerialName("product_name") val productName: String,
    val urgency: String = "",
    @SerialName("current_stock") val currentStock: Long = 0,
    @SerialName("avg_daily_velocity") val avgDailyVelocity: Double = 0.0,
    @SerialName("days_until_stockout") val daysUntilStockout: Int = 0,
    @SerialName("reorder_quantity") val reorderQuantity: Long = 0,
    val status: String = "",
    @SerialName("created_at") val createdAt: String = "",
)

@Serializable
data class ReplenishmentInsightsResponse(
    val insights: List<ReplenishmentInsight> = emptyList(),
    val data: List<ReplenishmentInsight> = emptyList(),
) {
    fun resolved(): List<ReplenishmentInsight> =
        insights.ifEmpty { data }
}

@Serializable
data class DispatchSettingsResponse(
    @SerialName("warehouse_id") val warehouseId: String = "",
    @SerialName("auto_dispatch_enabled") val autoDispatchEnabled: Boolean = false,
)

@Serializable
data class DispatchSettingsPatchRequest(
    @SerialName("auto_dispatch_enabled") val autoDispatchEnabled: Boolean,
)

@Serializable
data class ReplenishmentInsightActionResponse(
    @SerialName("insight_id") val insightId: String,
    val status: String,
    @SerialName("transfer_id") val transferId: String? = null,
)

@Serializable
data class OpsFinancialsResponse(
    @SerialName("warehouse_id") val warehouseId: String,
    val period: String = "",
    val currency: String = "",
    @SerialName("total_revenue") val totalRevenue: Long = 0,
    @SerialName("completed_orders") val completedOrders: Long = 0,
    @SerialName("avg_order_value") val avgOrderValue: Long = 0,
    @SerialName("platform_fee") val platformFee: Long = 0,
    @SerialName("net_payout") val netPayout: Long = 0,
    @SerialName("cash_pending") val cashPending: Long = 0,
    @SerialName("cash_collected") val cashCollected: Long = 0,
)

@Serializable
data class WarehouseOrderMutationRequest(
    val reason: String? = null,
)

@Serializable
data class WarehouseOrderMutationResponse(
    @SerialName("order_id") val orderId: String,
    val status: String,
)

@Serializable
data class RouteCoordinateWire(
    val lat: Double,
    val lng: Double,
)

@Serializable
data class RouteGeometryWire(
    @SerialName("route_id") val routeId: String? = null,
    @SerialName("encoded_polyline") val encodedPolyline: String? = null,
    val coordinates: List<RouteCoordinateWire> = emptyList(),
    val source: String = "",
    @SerialName("stop_count") val stopCount: Int? = null,
)

@Serializable
data class WarehouseDriverLocationWire(
    @SerialName("driver_id") val driverId: String,
    @SerialName("supplier_id") val supplierId: String? = null,
    val lat: Double = 0.0,
    val lng: Double = 0.0,
    val latitude: Double = 0.0,
    val longitude: Double = 0.0,
    @SerialName("reported_at") val reportedAt: String = "",
    @SerialName("received_at") val receivedAt: String = "",
    @SerialName("stale_after_seconds") val staleAfterSeconds: Int = 0,
)

@Serializable
data class WarehouseFleetLiveRoute(
    @SerialName("manifest_id") val manifestId: String,
    @SerialName("route_id") val routeId: String,
    @SerialName("driver_id") val driverId: String,
    @SerialName("driver_name") val driverName: String? = null,
    @SerialName("manifest_state") val manifestState: String,
    @SerialName("route_geometry") val routeGeometry: RouteGeometryWire? = null,
    @SerialName("driver_location") val driverLocation: WarehouseDriverLocationWire? = null,
    @SerialName("live_location_available") val liveLocationAvailable: Boolean = false,
    @SerialName("location_stale") val locationStale: Boolean? = null,
)

@Serializable
data class WarehouseFleetLiveMapResponse(
    val routes: List<WarehouseFleetLiveRoute> = emptyList(),
    @SerialName("warehouse_id") val warehouseId: String = "",
    @SerialName("fetched_at") val fetchedAt: String = "",
)

@Serializable
data class BroadcastTemplate(
    val id: String,
    val category: String = "",
    val title: String,
    val body: String,
    @SerialName("default_role") val defaultRole: String = "DRIVER",
    val scope: String = "warehouse",
    val source: String? = null,
    @SerialName("warehouse_id") val warehouseId: String? = null,
    @SerialName("placeholder_keys") val placeholderKeys: List<String>? = null,
)

@Serializable
data class BroadcastTemplatesResponse(
    val templates: List<BroadcastTemplate> = emptyList(),
)

@Serializable
data class WarehouseBroadcastRequest(
    val title: String,
    val body: String,
    val role: String? = null,
)

@Serializable
data class WarehouseBroadcastResponse(
    val status: String = "",
    @SerialName("warehouse_id") val warehouseId: String = "",
    @SerialName("supplier_id") val supplierId: String = "",
)

@Serializable
data class WarehouseBroadcastTemplateCreateRequest(
    val title: String,
    val body: String,
    @SerialName("default_role") val defaultRole: String? = null,
    val category: String? = null,
)

@Serializable
data class BroadcastTemplateDeleteResponse(
    val status: String = "",
    @SerialName("template_id") val templateId: String = "",
)

@Serializable
data class RetailerOverridePreview(
    @SerialName("retailers_on_sku_count") val retailersOnSkuCount: Int = 0,
    @SerialName("active_override_count") val activeOverrideCount: Int = 0,
    @SerialName("catalog_list_price") val catalogListPrice: Long = 0,
    @SerialName("margin_delta_per_unit") val marginDeltaPerUnit: Long = 0,
    @SerialName("margin_estimate_label") val marginEstimateLabel: String = "",
    @SerialName("affected_retailer_ids") val affectedRetailerIds: List<String>? = null,
    @SerialName("read_only") val readOnly: Boolean? = null,
)

@Serializable
data class RetailerOverridePreviewRequest(
    @SerialName("retailer_id") val retailerId: String? = null,
    @SerialName("product_id") val productId: String? = null,
    @SerialName("sku_id") val skuId: String? = null,
    @SerialName("proposed_price") val proposedPrice: Long,
)

File: pegasusX/apps/warehouse-app-ios/WarehouseApp/Models/WarehouseOpsModels.swift
import Foundation

struct EmergencyTransferRequest: Encodable {
    let totalVolumeVu: Double
    let notes: String?

    enum CodingKeys: String, CodingKey {
        case totalVolumeVu = "total_volume_vu"
        case notes
    }
}

struct ForceReceiveRequest: Encodable {
    let factoryId: String?
    let totalVolumeVu: Double
    let notes: String?

    enum CodingKeys: String, CodingKey {
        case factoryId = "factory_id"
        case totalVolumeVu = "total_volume_vu"
        case notes
    }
}

struct TransferMutationResponse: Decodable {
    let transferId: String
    let state: String
    let notes: String?

    enum CodingKeys: String, CodingKey {
        case transferId = "transfer_id"
        case state, notes
    }
}

struct ReplenishmentInsight: Decodable, Identifiable {
    let id: String
    let warehouseId: String
    let warehouseName: String
    let productId: String
    let productName: String
    let urgency: String
    let currentStock: Int64
    let avgDailyVelocity: Double
    let daysUntilStockout: Int
    let reorderQuantity: Int64
    let status: String
    let createdAt: String
    let reasonCode: String?
    let demandBreakdown: [String: AnyCodable]?

    enum CodingKeys: String, CodingKey {
        case id
        case warehouseId = "warehouse_id"
        case warehouseName = "warehouse_name"
        case productId = "product_id"
        case productName = "product_name"
        case urgency
        case currentStock = "current_stock"
        case avgDailyVelocity = "avg_daily_velocity"
        case daysUntilStockout = "days_until_stockout"
        case reorderQuantity = "reorder_quantity"
        case status
        case createdAt = "created_at"
        case reasonCode = "reason_code"
        case demandBreakdown = "demand_breakdown"
    }
}

struct ReplenishmentInsightsResponse: Decodable {
    let insights: [ReplenishmentInsight]
    let data: [ReplenishmentInsight]?

    var rows: [ReplenishmentInsight] {
        if !insights.isEmpty { return insights }
        return data ?? []
    }
}

struct ReplenishmentInsightActionResponse: Decodable {
    let insightId: String
    let status: String
    let transferId: String?

    enum CodingKeys: String, CodingKey {
        case insightId = "insight_id"
        case status
        case transferId = "transfer_id"
    }
}

struct DispatchSettingsResponse: Decodable {
    let warehouseId: String
    let autoDispatchEnabled: Bool

    enum CodingKeys: String, CodingKey {
        case warehouseId = "warehouse_id"
        case autoDispatchEnabled = "auto_dispatch_enabled"
    }
}

struct DispatchSettingsPatchRequest: Encodable {
    let autoDispatchEnabled: Bool

    enum CodingKeys: String, CodingKey {
        case autoDispatchEnabled = "auto_dispatch_enabled"
    }
}

struct OpsFinancialsResponse: Decodable {
    let warehouseId: String
    let period: String
    let currency: String
    let totalRevenue: Int64
    let completedOrders: Int64
    let avgOrderValue: Int64
    let platformFee: Int64
    let netPayout: Int64
    let cashPending: Int64
    let cashCollected: Int64

    enum CodingKeys: String, CodingKey {
        case warehouseId = "warehouse_id"
        case period, currency
        case totalRevenue = "total_revenue"
        case completedOrders = "completed_orders"
        case avgOrderValue = "avg_order_value"
        case platformFee = "platform_fee"
        case netPayout = "net_payout"
        case cashPending = "cash_pending"
        case cashCollected = "cash_collected"
    }
}

struct WarehouseOrderMutationRequest: Encodable {
    let reason: String?
}

struct WarehouseProposeDeliveryRequest: Encodable {
    let proposedDeliveryDate: String
    let reason: String

    enum CodingKeys: String, CodingKey {
        case proposedDeliveryDate = "proposed_delivery_date"
        case reason
    }
}

struct WarehouseOrderMutationResponse: Decodable {
    let orderId: String
    let status: String

    enum CodingKeys: String, CodingKey {
        case orderId = "order_id"
        case status
    }
}

struct RefreshTokenRequest: Encodable {
    let refreshToken: String

    enum CodingKeys: String, CodingKey {
        case refreshToken = "refresh_token"
    }
}

struct RouteCoordinateWire: Decodable {
    let lat: Double
    let lng: Double
}

struct RouteGeometryWire: Decodable {
    let routeId: String?
    let encodedPolyline: String?
    let coordinates: [RouteCoordinateWire]
    let source: String
    let stopCount: Int?

    enum CodingKeys: String, CodingKey {
        case routeId = "route_id"
        case encodedPolyline = "encoded_polyline"
        case coordinates, source
        case stopCount = "stop_count"
    }
}

struct WarehouseDriverLocationWire: Decodable {
    let driverId: String
    let supplierId: String?
    let lat: Double
    let lng: Double
    let latitude: Double
    let longitude: Double
    let reportedAt: String
    let receivedAt: String
    let staleAfterSeconds: Int

    enum CodingKeys: String, CodingKey {
        case driverId = "driver_id"
        case supplierId = "supplier_id"
        case lat, lng, latitude, longitude
        case reportedAt = "reported_at"
        case receivedAt = "received_at"
        case staleAfterSeconds = "stale_after_seconds"
    }

    var resolvedLatitude: Double { latitude != 0 ? latitude : lat }
    var resolvedLongitude: Double { longitude != 0 ? longitude : lng }
}

struct WarehouseFleetLiveRoute: Decodable, Identifiable {
    var id: String { manifestId }
    let manifestId: String
    let routeId: String
    let driverId: String
    let driverName: String?
    let manifestState: String
    let routeGeometry: RouteGeometryWire?
    let driverLocation: WarehouseDriverLocationWire?
    let liveLocationAvailable: Bool
    let locationStale: Bool?

    enum CodingKeys: String, CodingKey {
        case manifestId = "manifest_id"
        case routeId = "route_id"
        case driverId = "driver_id"
        case driverName = "driver_name"
        case manifestState = "manifest_state"
        case routeGeometry = "route_geometry"
        case driverLocation = "driver_location"
        case liveLocationAvailable = "live_location_available"
        case locationStale = "location_stale"
    }
}

struct WarehouseFleetLiveMapResponse: Decodable {
    let routes: [WarehouseFleetLiveRoute]
    let warehouseId: String
    let fetchedAt: String

    enum CodingKeys: String, CodingKey {
        case routes
        case warehouseId = "warehouse_id"
        case fetchedAt = "fetched_at"
    }
}

struct BroadcastTemplate: Decodable, Identifiable {
    let id: String
    let category: String
    let title: String
    let body: String
    let defaultRole: String
    let scope: String
    let source: String?
    let warehouseId: String?
    let placeholderKeys: [String]?

    enum CodingKeys: String, CodingKey {
        case id, category, title, body, scope, source
        case defaultRole = "default_role"
        case warehouseId = "warehouse_id"
        case placeholderKeys = "placeholder_keys"
    }
}

struct BroadcastTemplatesResponse: Decodable {
    let templates: [BroadcastTemplate]
}

struct WarehouseBroadcastRequest: Encodable {
    let title: String
    let body: String
    let role: String?
}

struct WarehouseBroadcastResponse: Decodable {
    let status: String
    let warehouseId: String
    let supplierId: String

    enum CodingKeys: String, CodingKey {
        case status
        case warehouseId = "warehouse_id"
        case supplierId = "supplier_id"
    }
}

struct WarehouseBroadcastTemplateCreateRequest: Encodable {
    let title: String
    let body: String
    let defaultRole: String?
    let category: String?

    enum CodingKeys: String, CodingKey {
        case title, body, category
        case defaultRole = "default_role"
    }
}

struct BroadcastTemplateDeleteResponse: Decodable {
    let status: String
    let templateId: String

    enum CodingKeys: String, CodingKey {
        case status
        case templateId = "template_id"
    }
}

struct RetailerOverridePreview: Decodable {
    let retailersOnSkuCount: Int
    let activeOverrideCount: Int
    let catalogListPrice: Int64
    let marginDeltaPerUnit: Int64
    let marginEstimateLabel: String
    let affectedRetailerIds: [String]?
    let readOnly: Bool?

    enum CodingKeys: String, CodingKey {
        case retailersOnSkuCount = "retailers_on_sku_count"
        case activeOverrideCount = "active_override_count"
        case catalogListPrice = "catalog_list_price"
        case marginDeltaPerUnit = "margin_delta_per_unit"
        case marginEstimateLabel = "margin_estimate_label"
        case affectedRetailerIds = "affected_retailer_ids"
        case readOnly = "read_only"
    }
}

struct RetailerOverridePreviewRequest: Encodable {
    let retailerId: String?
    let productId: String?
    let skuId: String?
    let proposedPrice: Int64

    enum CodingKeys: String, CodingKey {
        case retailerId = "retailer_id"
        case productId = "product_id"
        case skuId = "sku_id"
        case proposedPrice = "proposed_price"
    }
}

struct WarehousePulseEvent: Decodable, Identifiable {
    let id: String
    let kind: String
    let title: String
    let description: String?
    let occurredAt: String
    let deepLink: String?
    let orderId: String?
    let manifestId: String?

    enum CodingKeys: String, CodingKey {
        case id, kind, title, description
        case occurredAt = "occurred_at"
        case deepLink = "deep_link"
        case orderId = "order_id"
        case manifestId = "manifest_id"
    }
}

struct WarehousePulseResponse: Decodable {
    let events: [WarehousePulseEvent]
    let fetchedAt: String

    enum CodingKeys: String, CodingKey {
        case events
        case fetchedAt = "fetched_at"
    }
}

enum HandoffPulseSupport {
    private static let handoffKinds: Set<String> = [
        "PREORDER",
        "ORDER_ACCEPTED",
        "ORDER_DISPATCHED",
        "MANIFEST_SEALED",
        "MANIFEST_DISPATCHED",
        "DISPATCH",
    ]

    static func filter(_ events: [WarehousePulseEvent]) -> [WarehousePulseEvent] {
        events.filter(isHandoff)
    }

    private static func isHandoff(_ event: WarehousePulseEvent) -> Bool {
        let haystack = "\(event.kind) \(event.title)".uppercased()
        if handoffKinds.contains(event.kind.uppercased()) { return true }
        return haystack.range(of: "PREORDER|ACCEPT|DISPATCH|SEAL|MANIFEST", options: .regularExpression) != nil
    }
}

struct RecommendReassignRequest: Encodable {
    let orderId: String

    enum CodingKeys: String, CodingKey {
        case orderId = "order_id"
    }
}

struct TruckRecommendation: Decodable {
    let driverId: String
    let driverName: String
    let vehicleId: String
    let vehicleClass: String
    let licensePlate: String
    let score: Double
    let reason: String
    let toRoute: String

    enum CodingKeys: String, CodingKey {
        case driverId = "driver_id"
        case driverName = "driver_name"
        case vehicleId = "vehicle_id"
        case vehicleClass = "vehicle_class"
        case licensePlate = "license_plate"
        case score
        case reason
        case toRoute = "to_route"
    }
}

struct RecommendReassignResponse: Decodable {
    let orderId: String
    let retailerName: String
    let orderVolumeVu: Double
    let currentDriver: String
    let recommendations: [TruckRecommendation]

    enum CodingKeys: String, CodingKey {
        case orderId = "order_id"
        case retailerName = "retailer_name"
        case orderVolumeVu = "order_volume_vu"
        case currentDriver = "current_driver"
        case recommendations
    }
}

struct ReassignOrderRequest: Encodable {
    let orderId: String
    let toDriverId: String
    let reason: String
    let toManifestId: String?
    let isPartial: Bool

    enum CodingKeys: String, CodingKey {
        case orderId = "order_id"
        case toDriverId = "to_driver_id"
        case reason
        case toManifestId = "to_manifest_id"
        case isPartial = "is_partial"
    }
}
